File: python/algorithm.py
from abc import ABC, abstractmethod

from mapfmclient import Problem, Solution


class MapfAlgorithm(ABC):
    @abstractmethod
    def solve(self, problem: Problem) -> Solution:
        raise NotImplemented

    @property
    @abstractmethod
    def name(self) -> str:
        raise NotImplemented

    @property
    def version(self) -> str:
        return "0.0.1"

File: rust/src/algorithm.rs
use crate::problem::Problem;
use std::error::Error;
use std::fmt::Debug;
use crate::solution::Solution;


pub trait MapfmAlgorithm {
    type Error: Error + Debug;

    fn name() -> &'static str;
    fn version() -> &'static str {
        "0.0.1"
    }

    fn solve(problem: Problem) -> Result<Solution, Self::Error>;
}